### ISSUE TITLE
neuronx-cc 2.9 / 2.10 Compiler Crash on Softmax

### ISSUE THREAD

**Issue by xanderdunn** [aws-neuron/aws-neuron-sdk#740]
We have the following softmax graph:
[rust_hlo_softmax_parts.pb.zip](https://github.com/aws-neuron/aws-neuron-sdk/files/12601500/rust_hlo_softmax_parts.pb.zip)

```
debug_ir: HloModule xla_computation_ordered_wrapper, entry_computation_layout={(f32[3,5,7]{2,1,0}, f32[3,5,7]{2,1,0})->(f32[3,5,7]{2,1,0}, f32[3,5,7]{2,1,0})}

region_0.45 {
  Arg_0.42 = f32[] parameter(0)
  Arg_1.42 = f32[] parameter(1)
  ROOT maximum.44 = f32[] maximum(Arg_0.42, Arg_1.42)
}

region_1.49 {
  Arg_0.46 = f32[] parameter(0)
  Arg_1.46 = f32[] parameter(1)
  ROOT add.48 = f32[] add(Arg_0.46, Arg_1.46)
}

ENTRY main.52 {
  Arg_0.1 = f32[3,5,7]{2,1,0} parameter(0)
  constant.3 = f32[] constant(-inf)
  reduce.4 = f32[3,5]{1,0} reduce(Arg_0.1, constant.3), dimensions={2}, to_apply=region_0.45
  reshape.5 = f32[3,5,1]{2,1,0} reshape(reduce.4)
  reshape.6 = f32[3,5]{1,0} reshape(reshape.5)
  broadcast.7 = f32[3,5,7]{2,1,0} broadcast(reshape.6), dimensions={0,1}
  subtract.8 = f32[3,5,7]{2,1,0} subtract(Arg_0.1, broadcast.7)
  exponential.9 = f32[3,5,7]{2,1,0} exponential(subtract.8)
  constant.10 = f32[] constant(0)
  reduce.11 = f32[3,5]{1,0} reduce(exponential.9, constant.10), dimensions={2}, to_apply=region_1.49
  reshape.12 = f32[3,5,1]{2,1,0} reshape(reduce.11)
  reshape.13 = f32[3,5]{1,0} reshape(reshape.12)
  broadcast.14 = f32[3,5,7]{2,1,0} broadcast(reshape.13), dimensions={0,1}
  divide.15 = f32[3,5,7]{2,1,0} divide(exponential.9, broadcast.14)
  Arg_1.2 = f32[3,5,7]{2,1,0} parameter(1)
  divide.16 = f32[3,5,7]{2,1,0} divide(Arg_1.2, broadcast.14)
  negate.17 = f32[3,5,7]{2,1,0} negate(exponential.9)
  multiply.18 = f32[3,5,7]{2,1,0} multiply(broadcast.14, broadcast.14)
  divide.19 = f32[3,5,7]{2,1,0} divide(negate.17, multiply.18)
  multiply.20 = f32[3,5,7]{2,1,0} multiply(Arg_1.2, divide.19)
  constant.21 = f32[] constant(0)
  reduce.22 = f32[3,5]{1,0} reduce(multiply.20, constant.21), dimensions={2}, to_apply=region_1.49
  reshape.23 = f32[3,5,1]{2,1,0} reshape(reduce.22)
  reshape.24 = f32[3,5]{1,0} reshape(reshape.23)
  broadcast.25 = f32[3,5,7]{2,1,0} broadcast(reshape.24), dimensions={0,1}
  add.26 = f32[3,5,7]{2,1,0} add(divide.16, broadcast.25)
  multiply.27 = f32[3,5,7]{2,1,0} multiply(add.26, exponential.9)
  negate.28 = f32[3,5,7]{2,1,0} negate(multiply.27)
  constant.29 = f32[] constant(0)
  reduce.30 = f32[3,5]{1,0} reduce(negate.28, constant.29), dimensions={2}, to_apply=region_1.49
  reshape.31 = f32[3,5,1]{2,1,0} reshape(reduce.30)
  reshape.32 = f32[3,5]{1,0} reshape(reshape.31)
  broadcast.38 = f32[3,5,7]{2,1,0} broadcast(reshape.32), dimensions={0,1}
  reshape.33 = f32[3,5,1]{2,1,0} reshape(reduce.4)
  reshape.34 = f32[3,5]{1,0} reshape(reshape.33)
  broadcast.35 = f32[3,5,7]{2,1,0} broadcast(reshape.34), dimensions={0,1}
  compare.36 = pred[3,5,7]{2,1,0} compare(Arg_0.1, broadcast.35), direction=EQ
  convert.37 = f32[3,5,7]{2,1,0} convert(compare.36)
  multiply.39 = f32[3,5,7]{2,1,0} multiply(broadcast.38, convert.37)
  add.41 = f32[3,5,7]{2,1,0} add(multiply.27, multiply.39)
  ROOT tuple.50 = (f32[3,5,7]{2,1,0}, f32[3,5,7]{2,1,0}) tuple(divide.15, add.41)
}
```

The compiler crashes on it:
```
thread 'nn::tests::xla_softmax_parts' panicked at 'Script '"neuronx-cc" "compile" "/mnt/drive1/tmp/rust_hlo_softmax_parts.pb" "--framework=XLA" "--target=trn1" "--model-type=transformer" "--auto-cast=none" "--output" "/mnt/drive1/tmp/softmax_parts_13993278443889127085.neff"' failed:
Process Process-1:
Traceback (most recent call last):
  File "neuronxcc/driver/CommandDriver.py", line 345, in neuronxcc.driver.CommandDriver.CommandDriver.run_subcommand
  File "neuronxcc/driver/commands/CompileCommand.py", line 1127, in neuronxcc.driver.commands.CompileCommand.CompileCommand.run
  File "neuronxcc/driver/commands/CompileCommand.py", line 1086, in neuronxcc.driver.commands.CompileCommand.CompileCommand.runPipeline
  File "neuronxcc/driver/commands/CompileCommand.py", line 1103, in neuronxcc.driver.commands.CompileCommand.CompileCommand.runPipeline
  File "neuronxcc/driver/commands/CompileCommand.py", line 1106, in neuronxcc.driver.commands.CompileCommand.CompileCommand.runPipeline
  File "neuronxcc/driver/Job.py", line 324, in neuronxcc.driver.Job.SingleInputJob.run
  File "neuronxcc/driver/Job.py", line 350, in neuronxcc.driver.Job.SingleInputJob.runOnState
  File "neuronxcc/driver/Pipeline.py", line 30, in neuronxcc.driver.Pipeline.Pipeline.runSingleInput
  File "neuronxcc/driver/Job.py", line 324, in neuronxcc.driver.Job.SingleInputJob.run
  File "neuronxcc/driver/Job.py", line 350, in neuronxcc.driver.Job.SingleInputJob.runOnState
  File "neuronxcc/driver/jobs/Frontend.py", line 389, in neuronxcc.driver.jobs.Frontend.Frontend.runSingleInput
  File "neuronxcc/driver/jobs/Frontend.py", line 187, in neuronxcc.driver.jobs.Frontend.Frontend.runXLAFrontend
  File "neuronxcc/driver/jobs/Frontend.py", line 170, in neuronxcc.driver.jobs.Frontend.Frontend.runHlo2Penguin
neuronxcc.driver.Exceptions.CompilerInvalidInputException: ERROR: Failed command  /home/ubuntu/.local/lib/python3.8/site-packages/neuronxcc/starfish/bin/hlo2penguin --input /mnt/drive1/tmp/rust_hlo_softmax_parts.pb --out-dir ./ --output penguin.py --remat --max-costly-ops=2 --max-live-in-size=5 --max-remat-chain-size=10 --max-mem-multiple=1.8 --min-def-use-distance=500 --remat-policy=transformer --allow-same-pass-remat=true --split-abc --layers-per-module=1 --coalesce-all-gathers=false --coalesce-reduce-scatters=false --coalesce-all-reduces=false --emit-tensor-level-dropout-ops --emit-tensor-level-rng-ops --native-to-custom-softmax
------------
Reported stdout:
2023-09-13 16:04:56.615330: I tensorflow/core/util/port.cc:104] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
DEBUG: needsModular? No. macCnt 0
INFO: Switching to single-module compile. PrePartitionPipe skipped.

Histogram before graph level optimizations:
total HLO instructions: 41
             reshape        10  24.39% ################################################################
           broadcast         5  12.20% ################################
            constant         4   9.76% #########################
            multiply         4   9.76% #########################
              reduce         4   9.76% #########################
              divide         3   7.32% ###################
                 add         2   4.88% ############
              negate         2   4.88% ############
           parameter         2   4.88% ############
             compare         1   2.44% ######
             convert         1   2.44% ######
         exponential         1   2.44% ######
            subtract         1   2.44% ######
               tuple         1   2.44% ######


Histogram before graph level optimizations:
total HLO instructions: 41
             reshape        10  24.39% ################################################################
           broadcast         5  12.20% ################################
            constant         4   9.76% #########################
            multiply         4   9.76% #########################
              reduce         4   9.76% #########################
              divide         3   7.32% ###################
                 add         2   4.88% ############
              negate         2   4.88% ############
           parameter         2   4.88% ############
             compare         1   2.44% ######
             convert         1   2.44% ######
         exponential         1   2.44% ######
            subtract         1   2.44% ######
               tuple         1   2.44% ######

INFO: IoStatistics: total inputs: 2
INFO: IoStatistics: total outputs: 2
INFO: IoStatistics: total passthrough tensors: 0
INFO: IoStatistics: total outputs read from: 0
INFO: IoStatistics: total redundant outputs: 0
Replaced 0 dropout sequences with OffloadedDropout
INFO: HloMacCount has found 0
INFO: Traffic has found 1680
INFO: AIF 0.00

Histogram after graph level optimizations:
total HLO instructions: 29
           broadcast         6  20.69% ################################################################
              reduce         4  13.79% ##########################################
              divide         3  10.34% ################################
            multiply         3  10.34% ################################
                 add         2   6.90% #####################
            constant         2   6.90% #####################
              negate         2   6.90% #####################
           parameter         2   6.90% #####################
             compare         1   3.45% ##########
         exponential         1   3.45% ##########
              select         1   3.45% ##########
            subtract         1   3.45% ##########
               tuple         1   3.45% ##########

HLO Ops used in computation: add broadcast compare constant divide exponential multiply negate parameter reduce select subtract tuple
Number of native softmaxes detected and replaced: 0
Number of native softmaxes detected and replaced: 0

------------
Reported stderr:
None
------------
Import of the HLO graph into the Neuron Compiler has failed.
This may be caused by unsupported operators or an internal compiler error.
More details can be found in the error message(s) above.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "neuronxcc/driver/CommandDriver.py", line 352, in neuronxcc.driver.CommandDriver.CommandDriver.run_subcommand_in_process
  File "neuronxcc/driver/CommandDriver.py", line 347, in neuronxcc.driver.CommandDriver.CommandDriver.run_subcommand
  File "neuronxcc/driver/CommandDriver.py", line 113, in neuronxcc.driver.CommandDriver.handleError
  File "neuronxcc/driver/GlobalState.py", line 102, in neuronxcc.driver.GlobalState.FinalizeGlobalState
  File "neuronxcc/driver/GlobalState.py", line 82, in neuronxcc.driver.GlobalState._GlobalStateImpl.shutdown
  File "/usr/lib/python3.8/shutil.py", line 709, in rmtree
    onerror(os.lstat, path, sys.exc_info())
  File "/usr/lib/python3.8/shutil.py", line 707, in rmtree
    orig_st = os.lstat(path)
FileNotFoundError: [Errno 2] No such file or directory: '/home/ubuntu/dev/Kholinar/xla/neuronxcc-aaz2_gmb'
, stdout:
', xla/src/xla_utils.rs:49:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

This crash occurs both on neuronx-cc 2.9:
```
$ neuronx-cc --version
NeuronX Compiler version 2.9.0.40+07376825f

Python version 3.8.10
HWM version 2.9.0.2-f79d59e7b
NumPy version 1.21.6

Running on AMI ami-01257e71ecb2f431c
Running in region usw2-az4
```

And on neuronx-cc 2.10 beta:
```
$ neuronx-cc --version
NeuronX Compiler version 2.10.0.32+6d3873a1b

Python version 3.8.10
HWM version 2.10.0.5-7b1976adf
NumPy version 1.21.6

Running on AMI ami-01257e71ecb2f431c
Running in region usw2-az4
```

This is a regression. When I roll back to neuronx-cc 2.8.0.25+a3ad0f342, it successfully compiles. 

It compiles and runs on GPU and CPU using jax.

**Comment by shebbur-aws** [comment]
Thank you for reporting the issue. We are taking a look

**Comment by shebbur-aws** [comment]
We have a fix for this issue which will be available in one of the upcoming releases

**Comment by xanderdunn** [comment]
This is fixed in NRT SDK 2.15. 

```
$ neuronx-cc --version
NeuronX Compiler version 2.11.0.34+c5231f848

Python version 3.8.10
HWM version 2.11.0.2-e34678757
NumPy version 1.22.4

Running on AMI ami-01257e71ecb2f431c
Running in region usw2-az4
```
